scipy/integrate/_quadrature.py

```diff
@@ -353,31 +355,31 @@ def _basic_simpson(y, start, stop, x, dx, axis):
     slice0 = tupleset(slice_all, axis, slice(start, stop, step))
     slice1 = tupleset(slice_all, axis, slice(start+1, stop+1, step))
     slice2 = tupleset(slice_all, axis, slice(start+2, stop+2, step))
+    slice_early = tupleset(slice_all, axis, slice(None, -1))
+    slice_late = tupleset(slice_all, axis, slice(1, None))
 
     if x is None:  # Even-spaced Simpson's rule.
-        result = np.sum(y[slice0] + 4.0*y[slice1] + y[slice2], axis=axis)
+        result = xp.sum(y[slice0] + 4.0*y[slice1] + y[slice2], axis=axis)
         result *= dx / 3.0
     else:
         # Account for possibly different spacings.
         #    Simpson's rule changes a bit.
-        h = np.diff(x, axis=axis)
+        h = x[slice_late] - x[slice_early]
         sl0 = tupleset(slice_all, axis, slice(start, stop, step))
         sl1 = tupleset(slice_all, axis, slice(start+1, stop+1, step))
-        h0 = h[sl0].astype(float, copy=False)
-        h1 = h[sl1].astype(float, copy=False)
+        h0 = xp.astype(h[sl0], xp.float64, copy=False)
+        h1 = xp.astype(h[sl1], xp.float64, copy=False)
         hsum = h0 + h1
         hprod = h0 * h1
-        h0divh1 = np.true_divide(h0, h1, out=np.zeros_like(h0), where=h1 != 0)
-        tmp = hsum/6.0 * (y[slice0] *
-                          (2.0 - np.true_divide(1.0, h0divh1,
-                                                out=np.zeros_like(h0divh1),
-                                                where=h0divh1 != 0)) +
-                          y[slice1] * (hsum *
-                                       np.true_divide(hsum, hprod,
-                                                      out=np.zeros_like(hsum),
-                                                      where=hprod != 0)) +
-                          y[slice2] * (2.0 - h0divh1))
-        result = np.sum(tmp, axis=axis)
+        with np.errstate(divide='ignore', invalid='ignore'):
+            zero = xp.zeros_like(h0)
```

Comment:
Is creating this zero array and the several `xp.where`s  going to lead to performance issues for large h0/h1/hprod/hsum? Is a lazywhere better, or even fixing up afterwards.